src/compiler/counted_array.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 Mozilla Foundation
+// Original Copyright (c) 2016 Alex Burka, licensed to Mozilla Foundation under MIT license
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```

Comment:
The MIT license is very clear about this:

```
<snip>

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

// <snip>
```

I am not a lawyer, but to my understanding, it's not acceptable (at the very least morally) to change the copyright holder.

Reply:
> I have a hard time understanding why the 11 line solution is not sufficient, given the drive to make it easier to package while keeping things maintainable and reasonable level of chore?

The reason I didn't use your version of the macro is because your version doesn't handle scope correctly. 

![image](https://user-images.githubusercontent.com/3370875/185742541-102e1e9e-5322-4811-b780-bc684eff42f6.png)

Your 11 LOC version breaks on anything other than private static inputs (and we use public static, as evidenced by the above errors). I have added comment to the macro sections to help make it clear why each piece is necessary.

```rust
/**
 * Helper macro to create fixed-length arrays without specifying a static size
 */
#[macro_export]
macro_rules! counted_array {

    //CALL ENTRY

    // provided public array
    (pub $storage:ident $n:ident: [$t:ty; _] = [$($vals:expr),* $(,)*]) => {
        counted_array!(@parse 0usize, ($($vals),*) -> [] ((pub) $storage $n $t));
    };
    // provided private array
    ($storage:ident $n:ident: [$t:ty; _] = [$($vals:expr),* $(,)*]) => {
        counted_array!(@parse 0usize, ($($vals),*) -> [] (() $storage $n $t));
    };


    // RECURSIVE CALLS

    // tail call
    (@parse $size:expr, ($val:expr) -> [$($accs:expr),*] $thru:tt) => {
        counted_array!(@output $size + 1usize, [$($accs,)* $val] $thru);
    };
    // non-tail call
    (@parse $size:expr, ($val:expr, $($vals:expr),*) -> [$($accs:expr),*] $thru:tt) => {
        counted_array!(@parse $size + 1usize, ($($vals),*) -> [$($accs,)* $val] $thru);
    };

    //OUTPUT
    (@output $size:expr, $acc:tt (($($p:tt)*) $s:ident $n:ident $t:ty)) => {
        $($p)* $s $n: [$t; $size] = $acc;
    };
}
```

Reply:
> The MIT license is very clear about this:
> 
> ```
> <snip>
> 
> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
> 
> // <snip>
> ```
> 
> I am not a lawyer, but to my understanding, it's not acceptable (at the very least morally) to change the copyright holder.

Fixed.